src/main/java/bio/terra/service/dataset/flight/ingest/IngestValidateAzureRefsStep.java

```diff
@@ -0,0 +1,77 @@
+package bio.terra.service.dataset.flight.ingest;
+
+import bio.terra.common.Column;
+import bio.terra.common.Table;
+import bio.terra.model.BillingProfileModel;
+import bio.terra.model.TableDataType;
+import bio.terra.service.dataset.DatasetService;
+import bio.terra.service.filedata.azure.AzureSynapsePdao;
+import bio.terra.service.filedata.azure.tables.TableDirectoryDao;
+import bio.terra.service.profile.flight.ProfileMapKeys;
+import bio.terra.service.resourcemanagement.ResourceService;
+import bio.terra.service.resourcemanagement.azure.AzureAuthService;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.StepResult;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class IngestValidateAzureRefsStep extends IngestValidateRefsStep {
+
+  private final AzureAuthService azureAuthService;
+  private final ResourceService resourceService;
+  private final DatasetService datasetService;
+  private final AzureSynapsePdao azureSynapsePdao;
+  private final TableDirectoryDao tableDirectoryDao;
+
+  public IngestValidateAzureRefsStep(
+      ResourceService resourceService,
+      AzureAuthService azureAuthService,
+      DatasetService datasetService,
+      AzureSynapsePdao azureSynapsePdao,
+      TableDirectoryDao tableDirectoryDao) {
+    this.azureAuthService = azureAuthService;
+    this.resourceService = resourceService;
+    this.datasetService = datasetService;
+    this.azureSynapsePdao = azureSynapsePdao;
+    this.tableDirectoryDao = tableDirectoryDao;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException {
+    var workingMap = context.getWorkingMap();
+    var flightId = context.getFlightId();
+    var dataset = IngestUtils.getDataset(context, datasetService);
+
+    var billingProfile = workingMap.get(ProfileMapKeys.PROFILE_MODEL, BillingProfileModel.class);
+    var storageAccountResource =
+        resourceService.getOrCreateStorageAccount(dataset, billingProfile, flightId);
+
+    var tableServiceClient =
+        azureAuthService.getTableServiceClient(billingProfile, storageAccountResource);
+    Table table = IngestUtils.getDatasetTable(context, dataset);
+    var tableName = IngestUtils.getSynapseTableName(context.getFlightId());
+
+    // For each fileref column, scan the staging table and build an array of file ids
+    // Then probe the file system to validate that the file exists and is part
+    // of this dataset. We check all ids and return one complete error.
+
+    List<String> invalidRefIds =
+        table.getColumns().stream()
+            .map(Column::toSynapseColumn)
+            .filter(column -> column.getType() == TableDataType.FILEREF)
+            .flatMap(
+                column -> {
+                  List<String> refIdArray = azureSynapsePdao.getRefIds(tableName, column);
```

Comment:
Is it possible to have the same refid in more than one column? If so, it looks like you'll end up checking the same refid more than once.

Reply:
I've modified the synapse query to return `distinct` fileIds.